File: 42.trapping-rain-water.java
/*
 * @lc app=leetcode id=42 lang=java
 *
 * [42] Trapping Rain Water
 */

// @lc code=start
class Solution {
    public int trap(int[] height) {
        if (height == null || height.length < 3)
            return 0;

        int result = 0;
        int leftMax = Integer.MIN_VALUE; // highest bar from left
        int rightMax = Integer.MIN_VALUE; // highest bar from right

        // Key points:
        // all the bars between leftMax and rightMax do not matter
        for (int left = 0, right = height.length - 1; left <= right;) {
            leftMax = Math.max(leftMax, height[left]);
            rightMax = Math.max(rightMax, height[right]);

            // how much can current position trap depends on the shorter bar
            if (leftMax <= rightMax) {
                // DO NOT FORGET to subtract bar height of current position
                result += leftMax - height[left];
                left++;
            } else {
                result += rightMax - height[right];
                right--;
            }
        }
        return result;
    }
}
// @lc code=end

File: 42.trapping-rain-water.py
#
# @lc app=leetcode id=42 lang=python3
#
# [42] Trapping Rain Water
#

# @lc code=start
class Solution:
    def trap(self, height: List[int]) -> int:
        n = len(height)
        left, right = 0, n - 1      # 双指针初始指向两端
        left_max, right_max = 0, 0  # 记录到目前为止左右两端最高的“墙”
        ans = 0                     # 累计答案

        while left < right:
            if height[left] < height[right]:
                # 左边较低，以左边为限，能接的水 = left_max - height[left]
                if height[left] >= left_max:
                    left_max = height[left]  # 更新左侧最高
                else:
                    ans += left_max - height[left]
                left += 1
            else:
                # 右边较低，以右边为限，能接的水 = right_max - height[right]
                if height[right] >= right_max:
                    right_max = height[right]  # 更新右侧最高
                else:
                    ans += right_max - height[right]
                right -= 1

        return ans
        
# @lc code=end

